functions/go/set-namespace/transformer/namespace.go

```diff
@@ -0,0 +1,167 @@
+	namespacedResourcePattern = regexp.MustCompile(`\A([-.\w]*)/namespaces/([-.\w]*)/([-.\w]*)/([-.\w]*)\z`)
+	dependsOnKeyPattern       = func(group, kind, name string) string {
+		return fmt.Sprintf("%s/%s/%s", group, kind, name)
+	}
+)
+
+type Subject struct {
+	Kind      string `json:"kind,omitempty" yaml:"kind,omitempty"`
+	Name      string `json:"name,omitempty" yaml:"name,omitempty"`
+	ApiGroup  string `json:"apiGroup,omitempty" yaml:"apiGroup,omitempty"`
+	Namespace string `json:"namespace,omitempty" yaml:"namespace,omitempty"`
+}
+
+
+func SetNamespace(rl *fn.ResourceList) error {
+	tc := NamespaceTransformer{}
+	// Get "namespace" arguments from FunctionConfig
+	err := tc.Config(rl.FunctionConfig)
+	if err != nil {
+		rl.Results = append(rl.Results, fn.ErrorConfigObjectResult(err, rl.FunctionConfig))
+		return nil
+	}
+	// Update "namespace" to the proper resources.
+	tc.Transform(rl.Items)
+	var result *fn.Result
+	if len(tc.Errors) != 0 {
+		errMsg := strings.Join(tc.Errors, "\n")
+		result = fn.GeneralResult(errMsg, fn.Error)
+	} else {
+		result = fn.GeneralResult("namespace updated", fn.Info)
+	}
+	rl.Results = append(rl.Results, result)
+	return nil
+}
+
+type NamespaceTransformer struct {
+	Namespace    string
+	DependsOnMap map[string]bool
+	Errors       []string
+}
+
+func (p *NamespaceTransformer) Config(o *fn.KubeObject) error {
+	switch {
+	case o.IsGVK("v1", "ConfigMap"):
+		p.Namespace = o.GetStringOrDie("data", "namespace")
+		if p.Namespace == "" {
+			return fmt.Errorf("`data.namespace` should not be empty")
+		}
+	case o.IsGVK(fnConfigAPIVersion, fnConfigKind):
+		p.Namespace = o.GetStringOrDie("namespace")
+		if p.Namespace == "" {
+			return fmt.Errorf("`namespace` should not be empty")
+		}
+	case o.IsGVK("v1", "ConfigMap") && o.GetName() == "kptfile.kpt.dev":
+		p.Namespace = o.GetStringOrDie("data", "name")
+		if p.Namespace == "" {
+			return fmt.Errorf("`data.name` should not be empty")
+		}
+	default:
+		return fmt.Errorf("unknown functionConfig Kind=%v ApiVersion=%v, expect `%v` or `ConfigMap`",
+			o.GetKind(), o.GetAPIVersion(), fnConfigKind)
+	}
+	return nil
+}
+
+func (p *NamespaceTransformer) Transform(objects []*fn.KubeObject) {
+	p.SetDependsOnMap(objects)
+	for _, o := range objects {
+		switch {
+		case o.IsGVK("v1", "Namespace"):
+			o.SetName(p.Namespace)
+		case o.IsGVK("apiextensions.k8s.io/v1", "CustomResourceDefinition"):
+			err := o.Set(p.Namespace, "spec", "conversion", "webhook", "clientConfig", "service", "namespace")
+			if err != nil {
+				p.Errors = append(p.Errors, err.Error())
+			}
+		case o.IsGVK("apiregistration.k8s.io/v1", "APIService"):
+			err := o.Set(p.Namespace, "spec", "service", "namespace")
+			if err != nil {
+				p.Errors = append(p.Errors, err.Error())
+			}
+		case o.GetKind() == "ClusterRoleBinding" || o.GetKind() == "RoleBinding":
+			var subjects []*Subject
+			o.GetOrDie(&subjects, "subjects")
+			for _, s := range subjects {
+				if s.Name == "default" {
```

Comment:
As discussed in [this doc](https://docs.google.com/document/d/17rHGvJH7UTo9QaTa9w2kNJZvhkkCSQo4XrU-GWdTGQc/edit#bookmark=id.f4hvd131xsw0), there are 2 scenarios to update the namespace field:
- subject is a ServiceAccount and its name is "default"
- subject is a ServiceAccount and its name is not "default", but there is a ServiceAccount object that has the matching name.

It's a superset of what you currently have.

Reply:
Yeah, do you mind me making the change in the following PR. I have some other logic relevant to this need to add as well. Basically, the new set-namespace only expect a single namespace exists. It raises error if multiple namespace is detected. Thus, here the namespace will be updated if and only if it is a ServiceAccount kind and the existing namespace matches the found one.

Reply:
I'm OK with doing this logic in a followup PR.